File: android/app/src/main/kotlin/com/example/remote_control_app/models/Command.kt
import org.json.JSONObject

data class Command(
    val type: String,
    val startCoords: Pair,
    val endCoords: Pair
)

data class Pair(
    val x: Double,
    val y: Double
)

fun Command.toJson(): JSONObject {
    val jsonObject = JSONObject()
    jsonObject.put("type", type)
    jsonObject.put("startCoords", startCoords.toJson())
    jsonObject.put("endCoords", endCoords.toJson())
    return jsonObject
}

fun Pair.toJson(): JSONObject {
    val jsonObject = JSONObject()
    jsonObject.put("x", x)
    jsonObject.put("y", y)
    return jsonObject
}

fun JSONObject.toCommand(): Command {
    val type = getString("type")
    val startCoords = getJSONObject("startCoords").toPair()
    val endCoords = getJSONObject("endCoords").toPair()
    return Command(type, startCoords, endCoords)
}

fun JSONObject.toPair(): Pair {
    val x = getDouble("x")
    val y = getDouble("y")
    return Pair(x, y)
}

fun main() {
    // Create an instance of Command
    val command = Command("move", Pair(1.0, 2.0), Pair(3.0, 4.0))

    // Serialize to JSON
    val json = command.toJson()
    println("Serialized JSON: $json")

    // Deserialize from JSON
    val deserializedCommand = json.toCommand()
    println("Deserialized Command: $deserializedCommand")
}

File: lib/models/Command.dart
class Command {
  String type;
  Coords startCoords;
  Coords endCoords;

  Command(this.type, this.startCoords, this.endCoords);

  Map<String, dynamic> toJson() {
    return {
      'type': type,
      'startCoords': startCoords.toJson(),
      'endCoords': endCoords.toJson(),
    };
  }
}

class Coords {
  double x;
  double y;

  Coords(this.x, this.y);

  Map<String, dynamic> toJson() {
    return {
      'x': x,
      'y': y,
    };
  }
}